src/nccl_ofi_rdma.c

```diff
@@ -1622,98 +1601,109 @@ static int process_pending_reqs(nccl_net_ofi_rdma_ep_t *ep)
 					ret = ncclSystemError;
 					goto exit;
 				}
-				if (err_buffer.flags & FI_REMOTE_WRITE) {
-					req = get_req_from_imm_data(ep, err_buffer.data);
-					if (!req) {
-						NCCL_OFI_WARN("Unknown remote write error");
-						ret = ncclSystemError;
-						goto exit;
-					}
-				} else {
-					/* For all other operations, ctx should be a req */
-					if (!err_buffer.op_context) {
-						NCCL_OFI_WARN("Operation with NULL context completed with error!");
-						ret = ncclSystemError;
-						goto exit;
-					}
-					req = err_buffer.op_context;
-				}
-
-				NCCL_OFI_WARN("Request %p completed with error. RC: %d. Error: %s. Completed length: %ld, Request: %s",
-					      req,
-					      err_buffer.err,
-					      fi_cq_strerror(rail->cq,
-							     err_buffer.prov_errno,
-							     err_buffer.err_data, NULL, 0),
-					      (long)err_buffer.len,
-					      nccl_net_ofi_req_str(req));
-				set_request_state_to_error(req);
-
-				if (req->type == NCCL_OFI_RDMA_BOUNCE) {
-					/* A bounce buffer receive failed -- this is an internal error so bail out */
-					NCCL_OFI_WARN("Fatal: Bounce buffer recv completed with error");
+			} else {
+				/* For all other operations, ctx should be a req */
+				if (!err_buffer.op_context) {
+					NCCL_OFI_WARN("Operation with NULL context completed with error!");
 					ret = ncclSystemError;
 					goto exit;
 				}
-			} else if (rc == -FI_EAGAIN) {
-				/* No completions to process */
-				break;
-			} else {
-				NCCL_OFI_WARN("Unable to retrieve completion queue entries. RC: %zd, ERROR: %s",
-					      rc, fi_strerror(-rc));
+				req = err_buffer.op_context;
+			}
+
+			NCCL_OFI_WARN("Request %p completed with error. RC: %d. Error: %s. Completed length: %ld, Request: %s",
+				      req,
+				      err_buffer.err,
+				      fi_cq_strerror(rail->cq,
+						     err_buffer.prov_errno,
+						     err_buffer.err_data, NULL, 0),
+				      (long)err_buffer.len,
+				      nccl_net_ofi_req_str(req));
+			set_request_state_to_error(req);
+
+			if (req->type == NCCL_OFI_RDMA_BOUNCE) {
+				/* A bounce buffer receive failed -- this is an internal error so bail out */
+				NCCL_OFI_WARN("Fatal: Bounce buffer recv completed with error");
 				ret = ncclSystemError;
 				goto exit;
 			}
+		} else if (rc == -FI_EAGAIN) {
+			/* No completions to process */
+			break;
+		} else {
+			NCCL_OFI_WARN("Unable to retrieve completion queue entries. RC: %zd, ERROR: %s",
+				      rc, fi_strerror(-rc));
+			ret = ncclSystemError;
+			goto exit;
 		}
+	}
+
+exit:
+	return ret;
+}
+
+/*
+ * @brief	Process completion entries for the given completion quque.
+ *		This also updates several request fileds like size, status, etc
+ *
+ * @return	0, on success
+ *		error, on others
+ */
+static int ofi_process_cq(nccl_net_ofi_rdma_ep_t *ep)
+{
+	int ret;
+
+	for (int rail_id = 0; rail_id != ep->num_rails; ++rail_id) {
+		nccl_net_ofi_ep_rail_t *rail = get_rail(ep, rail_id);
 
+		ret = ofi_process_cq_rail(ep, rail);
```

Comment:
We are draining one rail's CQ after the other. Why is this preferable than instead doing a round-robin across rails?

Reply:
We're not changing that logic at all in this patch?

Reply:
No, I understand that you just moved things around and I am not saying we should do it here. Just wondering if it would make sense to do a round-robin.

Reply:
Yes, we should fiddle here. I played with that a bit, and it didn't make much difference, but that doesn't mean there isn't more room for investigation.